### ISSUE TITLE
Distributed training hangs at model checkpoint

### ISSUE THREAD

**Issue by Eralien** [PyTorchLightning/pytorch-lightning#10947]
## 🐛 Bug

I am trying to train a model in distributed training strategies using ddp, ddp_spawn and hovorod. All three methods hangs at the end of epochs that requires model checkpoint.

I have noticed that [manual-saving-with-strategies](https://pytorch-lightning.readthedocs.io/en/latest/common/weights_loading.html#manual-saving-with-strategies) has illustrated that with ddp model checkpoint should be used with either the trainer.save_checkpoint or by using `rank_zero_only()`. Both methods still hangs at the end of epochs requires model checkpoint.

The only successful way I found is by setting `pl.callbacks.ModelCheckpoint` with `monitor=None` and `save_top_k=0` which will immediately returns without actual saving actions. So I assume the hangs above is a deadlock between processes inside `callbacks.ModelCheckpoint` somewhere.

With other models I trained, this issue never exists. So I doubted that by saving the model I use somehow caused processes sync deadlocked. The model I trained has dynamic input shape and going thru several MLPs, and with a scatterND transformed the dynamic shape into a fixed-size pseudo image, and the rest is just normal yolo. The input "batch size" can varies between 8000~16000.

### To Reproduce

Sorry but to reproduce a distributed training on a BoringModel is troublesome, so I will give some pseudo code here.

The only way I can avoid the hangs is to skip any model saving:
```python
# trainer and checkpoint
def main(hparams: Namespace, config: Config):
    dataset = datamodule_factory(hparams, config)
    autoencoder = model_factory(hparams, config)
    checkpoint_callback = ModelCheckpoint(monitor=None,
                                          save_top_k=0,
                                          verbose=True,
                                          mode="min",
                                          every_n_epochs=10,
                                          save_on_train_epoch_end=False)
    trainer = pl.Trainer.from_argparse_args(hparams,
                                            callbacks=[checkpoint_callback])
    trainer.fit(autoencoder, dataset)

if __name__ == '__main__':
    main()
```

Monitoring metrics with `rank_zero_only` or `dist_sync`  hangs:
```python
# trainer part
checkpoint_callback = ModelCheckpoint(monitor="val/ave_obj",
                                          mode="min",
                                          every_n_epochs=10,
                                          save_on_train_epoch_end=False)

# validation part
def validation_step(self, batch, batch_idx):
    features, coors, targets = batch # dynamic sized features and coords
    pseudo_img = self._voxel_feat_extract(features, coors)
    outputs = self._backbone(pseudo_img, targets)
    ave_obj = (outputs[:, :, 4] > self._thresh).sum() / outputs.shape[0]
    self.log("val/ave_obj", ave_obj, rank_zero_only=True) # hangs
    self.log("val/ave_obj", ave_obj, dist_sync=True) # hangs
```

### Expected behavior

No hanging.

### Environment

* CUDA:
        - GPU:
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.3
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.2+cu111 / 1.10+cu113
        - pytorch-lightning: 1.5.4 / 1.5.3
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.12
        - version:           #171-Ubuntu SMP Fri Nov 5 11:55:11 UTC 2021

cc @tchaton @rohitgr7 @carmocca @awaelchli @ninginthecloud @jjenniferdai @justusschock @kaushikb11 @akihironitta @edward-io @ananthsub

**Comment by tchaton** [comment]
Hey @Eralien,

Would you be able to provide a reproducible script using the BoringModel ?

Here are some suggestions to investigate your problem:


* 1. Make sure you have the exact same number of batches on all ranks
* 2. Make sure you don't perform any distributed operation while saving a checkpoint

T.C

**Comment by Eralien** [comment]
Thanks @tchaton for you quick responce! Indeed I have never experienced such difficulties when using fixed batch size input. I will reproduce it on a BoringModel ASAP.

About your suggestions:

1. My model does not have the exact same number of batches on all ranks due to its nature, yet I can do some dirty tricks to make it be all the same. Should I remain batches the same on different steps? 
2. I am using the pl.callback.ModelCheckpoint. I assume it performs distributed operation by default (monitoring metrics on different ranks). So is there any way I can avoid distributed operation to achieve the same effect? Or do I have to write my own callback with tricks?

**Comment by Eralien** [comment]
[bug_report_model_#10947.ipynb](https://colab.research.google.com/drive/1qE29CVUleAcHNj43qSTAAwbZt-jjEjW2?usp=sharing)
Here's the BoringModel. I failed to run the code in colab with distributed mode, only on my local machine.

**Comment by carmocca** [comment]
Hi @Eralien 

I ran the following script (based off your colab link) on 4 GPUs, but did not see any hang.

```python
import os

from typing import Optional

import torch
from torch.utils.data import DataLoader, Dataset
import numpy as np
from numpy import random

import pytorch_lightning as pl
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.plugins import DDPPlugin
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, num_samples):
        self.len = num_samples
        self.data_size = [random.randint(2000, 20000) for _ in range(num_samples)]
        # self.data = torch.randn(num_samples, size)

    def __getitem__(self, index):
        return torch.randn(self.data_size[index], 8) * 255, torch.randint(100, (self.data_size[index], 2)), torch.ones(3, 10)
        # return self.data[index]

    def __len__(self):
        return self.len


def collate_fn_train(batches):
    features_batch = []
    coors_batch = []
    targets = []
    for i, (features, coors, target) in enumerate(batches):
        features_batch.append(features)
        coors[:, -1] = i
        coors_batch.append(coors)
        targets.append(target)
    return torch.cat(features_batch), torch.cat(coors_batch), torch.stack(targets)


class RandomDataModule(pl.LightningDataModule):
    def __init__(self, train_dataset, val_dataset, batch_size):
        super().__init__()
        self._train_dataset = train_dataset
        self._val_dataset = val_dataset
        self.batch_size = batch_size

    def train_dataloader(self):
        return DataLoader(self._train_dataset, batch_size=self.batch_size, collate_fn=collate_fn_train, num_workers=4)

    def val_dataloader(self):
        return DataLoader(self._val_dataset, batch_size=self.batch_size, collate_fn=collate_fn_train, num_workers=4)

    def prepare_data(self):
        pass

    def setup(self, stage: Optional[str] = None):
        pass

num_samples = 10000

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(8, 16)

    def forward(self, x):
        features, coors, targets = x
        y = self.layer(features)
        return y

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_val_data = RandomDataModule(RandomDataset(200), RandomDataset(20), batch_size=2)
    checkpoint_callback = ModelCheckpoint(monitor="valid_loss",
                                          mode="min",
                                          every_n_epochs=2,
                                          save_on_train_epoch_end=False)
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        strategy="ddp",
        accelerator="gpu",
        devices=4,
        num_sanity_val_steps=0,
        max_epochs=8,
        enable_model_summary=False,
        callbacks=[checkpoint_callback]
    )
    trainer.fit(model, train_val_data)

if __name__ == "__main__":
    run()
```

Can you verify that it still hangs for you on master?

**Comment by Eralien** [comment]
@carmocca Thanks. You're right, the sample does not hang. I found the wrong part where my code actually hangs. I will open another issue with corrected info.

**Comment by Eralien** [comment]
I found out I had been looking into a wrong direction. It probably have a similar reason (maybe?) like #11242 #9611 #11242 #11910

In #11910, as [qmpzzpmq](https://github.com/qmpzzpmq) mentioned [here](https://github.com/PyTorchLightning/pytorch-lightning/issues/11910#issuecomment-1055121784), I added both `limit_train_batches` and `limit_val_batches` as **int**  in command line options, and the random hanging is gone.

I do not understand why, but here's an observation. I printed out the length of train dataset and val dataset,  and in different nodes of the DDP, the dataset length are all different.
```
# I use three GPUs, 16 batch size
train length 33684, ~ 701.75 steps
train length 34113, ~ 710.68 steps
train length 34631, ~ 721.45 steps
val length 1701                                                                                                                                                                                                                                                          
val length 2059
val length 2311
```
It seems the steps in each rank are quite different.

What I did was manually calculate the minimal equivalent steps among all nodes, and manually add to `limit_train_batches` and `limit_val_batches`. If the limit number was any bigger, the hang would happen again.

**Comment by Eralien** [comment]
I did some further tests. The steps in different GPUs do lead to hang.
  
![image](https://user-images.githubusercontent.com/26458167/156703439-25aa7b6a-24fd-4202-a34e-749611440523.png)
rank 0 headed straight into validation loop, possibly causing the hang. 

Hang with all GPU 100% usage when NCCL was used as backend.
![image](https://user-images.githubusercontent.com/26458167/156703473-a84634b2-818b-49a9-98f9-024be4754722.png)

**Comment by Eralien** [comment]
Oh I see where the problem is. I introduced randomness somewhere in my dataset loading function, which was not a problem under PyTorch paradigm. 

In PyTorch the rank 0 process would only generate randomness for once, each yield one sample of train/val dataset number, and the Distributed Sampler would distribute that one sample of train/val number into different nodes, thus each node has the same set of steps in one epoch.

But PL spawn several processes that call my dataset loading function separately, where my random shuffle yields different length of steps! And sometimes the hang would just gone due to luck...

**Comment by 834799106** [comment]
Hi, @Eralien , this happened to me too, so your solution is to set shuffle to False?

**Comment by Eralien** [comment]
> Hi, @Eralien , this happened to me too, so your solution is to set shuffle to False?

Hi, I have two solutions. The first is indeed to set shuffle to False. The second is to set a static random seed for the shuffle part, to ensure reproducibility among different nodes.

**Comment by arlofaria** [comment]
> Oh I see where the problem is. I introduced randomness somewhere in my dataset loading factory function, 
> ...
> But PL spawns several processes that would call my dataset loading function separately, where my random shuffle yields different length of steps! And sometimes the hang would just gone due to luck...

@Eralien, I had the exact same problem/solution, and would note that it was an especially insidious bug to figure out!

In my case, each of the replicas would seed a custom `BatchSampler` with the same seed for the first epoch, but not re-seed it for later epochs; this was the bug, which was easily fixed once it was identified.

However, it was hard to spot the bug since you had to wait until one full epoch had finished, and even then it was non-deterministic whether the second or later epochs would stall.  Since each replica processed different data in the first epoch, their RNG states would diverge by the second epoch (I'm not certain where else the RNG was used, but possibly for dithering audio waveforms).  The problem, as you noted, was that my `BatchSampler` might produce a different number of batches for each replica, depending on how the dataset was shuffled.  However, that difference only mattered insofar as the number of batches differed when subjected to integer division by the number of replicas in a subsequent custom `DistributedSampler`.  To some extent, that would "hide" the bug especially if training with a large number of replicas.

So this specific bug would be most probable to appear after the non-initial epoch of experiments using a combination of large dataset (i.e. large number of batches and higher variance in the different number of batches on each replica) as well as a relatively small number of replicas (higher likelihood that the integer division would result in differing number of distributed batches).  In my case, I was able to most reliably reproduce the bug when limiting the jobs to just 2 replicas.

**Comment by ShaojieJiang** [comment]
> > Hi, @Eralien , this happened to me too, so your solution is to set shuffle to False?
> 
> Hi, I have two solutions. The first is indeed to set shuffle to False. The second is to set a static random seed for the shuffle part, to ensure reproducibility among different nodes.

Thanks @Eralien , this also solved my problem.

**Comment by zshuyinggg** [comment]
> 

Thank you!! This is very helpful! It solved my problem and saved my day!

**Comment by Liuykang** [comment]
> > Hi, @Eralien , this happened to me too, so your solution is to set shuffle to False?
> 
> Hi, I have two solutions. The first is indeed to set shuffle to False. The second is to set a static random seed for the shuffle part, to ensure reproducibility among different nodes.

Tanks, this solution is not valid for my situation. My program was hung at the end of the first epoch and the GPU usage rate reached 100%.
`pl.seed_everything(seed)`
trainer = pl.Trainer(
        accelerator="gpu",
        devices=[0, 1],
        benchmark=True,
        deterministic=False,
        max_epochs=10,
        callbacks=callbacks,
        logger=tb_logger,
        num_sanity_val_steps=0, 
        limit_train_batches=0.01,
        limit_val_batches=0.1,
        enable_progress_bar=True,
        strategy=DDPStrategy(find_unused_parameters=True),
    )
CUDA 11.8
python 3.8
torch: 2.0.1
pytorch-lightning 2.0.4
